### ISSUE TITLE
Can't produce zip with STORE method (uncompressed data)

### ISSUE THREAD

**Issue by darekrusin** [rubyzip/rubyzip#286]
When I pack a zip using WinRar for example and select the "store" method for an uncompressed archive, the algorithm used is reported as "store" and the file size is only a few hundred bytes larger than the compressed data.

But using rubyzip, even when I select Zlib::NO_COMPRESSION something weird happens:
- the file indeed has no compression (I can see parts of the original data readable inside), but the resulting zip size is larger, it gains about 8 KB on a 60 MB file
- the file being zipped is not stored as a raw blog inside the zip but it's chunked with some kind of checksums mixed in
- the algorithm reported is "deflate" instead of "store"

This is a problem for situation where a truly uncompressed zip is needed, with a raw binary version of the original file inside.

Am I doing something wrong? Is there a way to force the gem to work this way, or is this a bug?

**Comment by julik** [comment]
`Zlib::NO_COMPRESSION` is a setting in Zlib that is probably not usable with ZIPs. To use the stored mode you have to use the STORED mode in RubyZip. We use it in our streaming Zip archiving library but I do not remember whether RubyZip exposes it natively. I think it does.
TL;DR - `Zlib::NO_COMPRESSION` is actually _also_ a sort of deflate, so what you are seeing makes sense.

**Comment by hainesr** [comment]
I've forced a STORE like this in the past: https://github.com/myGrid/ruby-zip-container/blob/master/lib/zip-container/file.rb#L79

I'm pretty sure that does what you want.

**Comment by atz** [comment]
Same problem, as affecting our project here:
https://github.com/sul-dlss/preservation_catalog/issues/737#issuecomment-388504059

See the expected output (like `zip -vr0X -s 10g ...`):
```
$ unzip -lv example.zip
Archive:  example.zip
 Length   Method    Size  Cmpr    Date    Time   CRC-32   Name
--------  ------  ------- ---- ---------- ----- --------  ----
       0  Stored        0   0% 04-20-2018 12:06 00000000  jj925bx9565/v0001/
       0  Stored        0   0% 04-20-2018 12:06 00000000  jj925bx9565/v0001/data/
       0  Stored        0   0% 04-20-2018 12:06 00000000  jj925bx9565/v0001/data/content/
 6508496  Stored  6508496   0% 04-20-2018 12:06 ac341bd4  jj925bx9565/v0001/data/content/PC0170_s3_4th_of_July_2010-07-04_095432_0003.jpg
       0  Stored        0   0% 04-20-2018 12:06 00000000  jj925bx9565/v0001/data/metadata/
     873  Stored      873   0% 04-20-2018 12:06 960c6665  jj925bx9565/v0001/data/metadata/contentMetadata.xml
     345  Stored      345   0% 04-20-2018 12:06 be89e932  jj925bx9565/v0001/data/metadata/descMetadata.xml
     645  Stored      645   0% 04-20-2018 12:06 8d12b9d6  jj925bx9565/v0001/data/metadata/identityMetadata.xml
     263  Stored      263   0% 04-20-2018 12:06 7d1a7d24  jj925bx9565/v0001/data/metadata/provenanceMetadata.xml
     721  Stored      721   0% 04-20-2018 12:06 d0610866  jj925bx9565/v0001/data/metadata/relationshipMetadata.xml
     684  Stored      684   0% 04-20-2018 12:06 c07d0016  jj925bx9565/v0001/data/metadata/rightsMetadata.xml
    4490  Stored     4490   0% 04-20-2018 12:06 58e11294  jj925bx9565/v0001/data/metadata/technicalMetadata.xml
     134  Stored      134   0% 04-20-2018 12:06 72983849  jj925bx9565/v0001/data/metadata/versionMetadata.xml
    4133  Stored     4133   0% 04-20-2018 12:06 ce5990ca  jj925bx9565/v0001/data/metadata/workflows.xml
       0  Stored        0   0% 04-20-2018 12:06 00000000  jj925bx9565/v0001/manifests/
    4125  Stored     4125   0% 04-20-2018 12:06 22fd4501  jj925bx9565/v0001/manifests/fileInventoryDifference.xml
    1614  Stored     1614   0% 04-20-2018 12:06 44d2eaf3  jj925bx9565/v0001/manifests/manifestInventory.xml
    3118  Stored     3118   0% 04-20-2018 12:06 17bf12ff  jj925bx9565/v0001/manifests/signatureCatalog.xml
    3455  Stored     3455   0% 04-20-2018 12:06 c79865fb  jj925bx9565/v0001/manifests/versionAdditions.xml
    3527  Stored     3527   0% 04-20-2018 12:06 7e525d93  jj925bx9565/v0001/manifests/versionInventory.xml
--------          -------  ---                            -------
 6536623          6536623   0%                            20 files
```
Actual output from `rubyzip`:
```
unzip -lv rubyzip.zip 
Archive:  rubyzip.zip
 Length   Method    Size  Cmpr    Date    Time   CRC-32   Name
--------  ------  ------- ---- ---------- ----- --------  ----
       0  Stored        0   0% 05-11-2018 16:32 00000000  jj925bx9565/v0001/
       0  Stored        0   0% 05-11-2018 16:32 00000000  jj925bx9565/v0001/data/
       0  Stored        0   0% 05-11-2018 16:32 00000000  jj925bx9565/v0001/data/content/
 6508496  Defl:N  6509486   0% 05-11-2018 16:32 ac341bd4  jj925bx9565/v0001/data/content/PC0170_s3_4th_of_July_2010-07-04_095432_0003.jpg
       0  Stored        0   0% 05-11-2018 16:32 00000000  jj925bx9565/v0001/data/metadata/
     873  Defl:N      878  -1% 05-11-2018 16:32 960c6665  jj925bx9565/v0001/data/metadata/contentMetadata.xml
     345  Defl:N      350  -1% 05-11-2018 16:32 be89e932  jj925bx9565/v0001/data/metadata/descMetadata.xml
     645  Defl:N      650  -1% 05-11-2018 16:32 8d12b9d6  jj925bx9565/v0001/data/metadata/identityMetadata.xml
     263  Defl:N      268  -2% 05-11-2018 16:32 7d1a7d24  jj925bx9565/v0001/data/metadata/provenanceMetadata.xml
     721  Defl:N      726  -1% 05-11-2018 16:32 d0610866  jj925bx9565/v0001/data/metadata/relationshipMetadata.xml
     684  Defl:N      689  -1% 05-11-2018 16:32 c07d0016  jj925bx9565/v0001/data/metadata/rightsMetadata.xml
    4490  Defl:N     4495  -0% 05-11-2018 16:32 58e11294  jj925bx9565/v0001/data/metadata/technicalMetadata.xml
     134  Defl:N      139  -4% 05-11-2018 16:32 72983849  jj925bx9565/v0001/data/metadata/versionMetadata.xml
    4133  Defl:N     4138  -0% 05-11-2018 16:32 ce5990ca  jj925bx9565/v0001/data/metadata/workflows.xml
       0  Stored        0   0% 05-11-2018 16:32 00000000  jj925bx9565/v0001/manifests/
    4125  Defl:N     4130  -0% 05-11-2018 16:32 22fd4501  jj925bx9565/v0001/manifests/fileInventoryDifference.xml
    1614  Defl:N     1619  -0% 05-11-2018 16:32 44d2eaf3  jj925bx9565/v0001/manifests/manifestInventory.xml
    3118  Defl:N     3123  -0% 05-11-2018 16:32 17bf12ff  jj925bx9565/v0001/manifests/signatureCatalog.xml
    3455  Defl:N     3460  -0% 05-11-2018 16:32 c79865fb  jj925bx9565/v0001/manifests/versionAdditions.xml
    3527  Defl:N     3532  -0% 05-11-2018 16:32 7e525d93  jj925bx9565/v0001/manifests/versionInventory.xml
--------          -------  ---                            -------
 6536623          6537683   0%                            20 files

```
This is because `Zip::Entry.new` doesn't honor `Zip.default_compression`, and just defaults to deflate:
https://github.com/rubyzip/rubyzip/blob/41fcf0a26edf89cc8cf3f42109ac56f99e56d2d2/lib/zip/entry.rb#L66

**Comment by hainesr** [comment]
I'm looking at this and trying to work out what is going on.

Firstly, this isn't obvious and I had to do some digging to figure this out, but `Zip.default_compression` is for setting the *level* of compression, not the *method*. So setting `@compression_method` to `Zip.default_compression` in Entry wouldn't be right. That said, the Deflate compression method should be able to do pass-through compression (i.e. none) and `Zip.default_compression` certainly should be able to be set to `Zlib::NO_COMPRESSION` and that be honoured.

It looks to me like you can't use `Zip:File.add` to add a file with Store compression with the API as it is at the moment. I've used the workaround in my comment above before, and this definitely works, but seems ungainly.

I'm going to see if I can figure out why level 0 compression (i.e. `Zlib::NO_COMPRESSION`) isn't working for Deflate.
